cmd/gazelle/update-repos.go

```diff
@@ -245,23 +250,37 @@ func updateImportPaths(c *updateReposConfig, f *rule.File, kinds map[string]rule
 			return err
 		}
 	}
-	merger.MergeFile(f, nil, genRules, merger.PreResolve, kinds)
+	rulesByFile := repo.FilterRulesByFile(repos, genRules, destFile)
+	for f, rules := range rulesByFile {
+		merger.MergeFile(f, nil, rules, merger.PreResolve, kinds)
+		if err := f.Save(f.Path); err != nil {
+			return fmt.Errorf("error writing %q: %v", f.Path, err)
+		}
+	}
 	return nil
 }
 
-func importFromLockFile(c *updateReposConfig, f *rule.File, kinds map[string]rule.KindInfo) error {
-	rs := repo.ListRepositories(f)
-	rc, cleanupRc := repo.NewRemoteCache(rs)
+func importFromLockFile(c *updateReposConfig, workspace *rule.File, destFile *rule.File, kinds map[string]rule.KindInfo) error {
+	repos, err := repo.ListRepositories(workspace)
+	if err != nil {
+		return err
+	}
+	rc, cleanupRc := repo.NewRemoteCache(repos)
 	defer cleanupRc()
 	genRules, err := repo.ImportRepoRules(c.lockFilename, rc)
 	if err != nil {
 		return err
 	}
-	for i := range genRules {
-		applyBuildAttributes(c, genRules[i])
+	rulesByFile := repo.FilterRulesByFile(repos, genRules, destFile)
+	for f, rules := range rulesByFile {
+		for i := range rules {
+			applyBuildAttributes(c, rules[i])
+		}
+		merger.MergeFile(f, nil, rules, merger.PreResolve, kinds)
+		if err := f.Save(f.Path); err != nil {
```

Comment:
What happens if the same file contains multiple macros? It seems like earlier changes would be overwritten here.

Reply:
My latest change handles this case in `MergeRules` by calling `rule.SyncMacroFile`. It is not the greatest solution, so I am open to other ideas.